src/service/crypto-verifier-impl.js

```diff
@@ -0,0 +1,198 @@
+
+            .then(publicKey => crossCrypto.verify({
+              name: 'RSASSA-PKCS1-v1_5',
+              hash: 'SHA-256',
+            }, publicKey, new Uint8Array(signature, 5), signedData))
+
+            .catch(err => {
+              dev.error(TAG_, 'Error while verifying:', err);
+              throw err;
+            });
+        };
+      };
+      return Promise.resolve(false);
+    });
+  }
+
+  /**
+   * Is this service actually available? For now, we use browser native
+   * crypto. So if that is not available, then this service is not available.
+   * @return {boolean}
+   */
+  isAvailable() {
+    return !!crossCrypto;
+  }
+}
+
+/**
+ * Appends 4-byte endian data's length to the data itself.
+ * @param {!Uint8Array} the data.
+ * @return {!Uint8Array} the appended 4-byte endian data's length together with
+ *     the data itself.
+ */
+function lenPrefix(data) {
+  if (data == null || data.length == 0) {
+    return new Uint8Array([0x00, 0x00, 0x00, 0x00]);
+  }
+  if (data.length >= Math.pow(2, 32)) {
+    throw Error('Data\'s length can not exceed 2^32 - 1.');
+  }
+
+  const res = new Uint8Array(4 + data.length);
+  res[0] = (data.length >> 24) & 0xff;
+  res[1] = (data.length >> 16) & 0xff;
+  res[2] = (data.length >> 8) & 0xff;
+  res[3] = data.length & 0xff;
+  res.set(data, 4);
+  return res;
+}
+
+function asciiToUint8Array(str) {
+  const chars = new Uint8Array(str.length);
+  for (let i = 0; i < str.length; ++i) {
+    chars[i] = str.charCodeAt(i);
+  }
+  return chars;
+}
+
+/**
+ * @param {{e: !Uint8Array, n:!Uint8Array, keyHash: !Uint8Array }} the RSA
+ *     public key. If keyHash is null, this function will compute it.
+ * @return {!Promise<!ArrayBuffer>} the hash of RSA public key.
+ */
+function hashRSAPubKey(rsaPubKey) {
+  if (rsaPubKey['keyHash'] == null) {
+    const lenMod = lenPrefix(rsaPubKey['n']);
+    const lenPubExp = lenPrefix(rsaPubKey['e']);
+    const data = new Uint8Array(lenMod.length + lenPubExp.length);
+    data.set(lenMod);
+    data.set(lenPubExp, lenMod.length);
+    return crossCrypto.digest(
+      {
+        // The list of RSA public keys are not under attacker's control,
+        // so a collision would not help.
+        name: 'SHA-1',
+      },
+      data);
+  } else {
+    return Promise.resolve(rsaPubKey['keyHash']);
+  }
+}
+
+/**
+ * @param {{e: !Uint8Array, n:!Uint8Array, keyHash: !Uint8Array }} the RSA
+ *     public key. If keyHash is null, this function will compute it.
+ * @return {!Promise<!ArrayBuffer>} the hash of RSA public key.
+ */
+function hashRSAPubKeys(rsaPubKeys) {
+  return Promise.all(rsaPubKeys.map(rsaPubKey => hashRSAPubKey(rsaPubKey)));
+}
+
+/**
+ * Compare the hash field of the signature to keyHash.
+ * Note that signature has a one-byte version, followed by 4-byte hash.
+ * @param {!Uint8Array} signature
+ * @param {!Uint8Array} keyHash
+ * @return {boolean} signature[1..5] == keyHash
+ */
+function hashesEqual(signature, keyHash) {
```

Comment:
Is checking the length of the hash really all that's needed to verify that this signature was generated using this RSA key?